### ISSUE TITLE
Any reason not to accept ambiguous bases? 

### ISSUE THREAD

**Issue by 7PintsOfCherryGarcia** [MikkelSchubert/adapterremoval#67]
Kind of annoying having to remove them in order for AR to work.

**Comment by MikkelSchubert** [comment]
Hi,

You mean bases such as `W`, `S`, `M`, and so on, right? The reason why such bases are not accepted is that they are generally not expected in FASTQ reads and because handling them (correctly) makes everything a lot more complicated (and slower). For that reason it unlikely that (proper) support will be added to AdapterRemoval. 

However, I could probably add an option to simply mask ambiguous bases in the input (i.e. replace them with 'N'). Would such an option be of interest?

`N` bases on the other hand should be accepted everywhere except in demultiplexing barcodes. Support for 'N' (and potentially other ambiguous bases) in barcodes is not out of the question, but it'd take more work than an option to simply mask ambiguous bases.

Best,
Mikkel

**Comment by 7PintsOfCherryGarcia** [comment]
Hi, 

Masking of non ACGTN characters would be amazing. I guess it is not a complicated change? 
Something in:

```
void
fastq::post_process(const fastq_encoding& encoding)
{
  for (char& nuc : m_sequence) {
    // Fast ASCII letter uppercase
    switch (nuc &= 0xDF) {
      case 'A':
      case 'C':
      case 'G':
      case 'T':
      case 'N':
        break;

      default:
        throw fastq_error("invalid character in FASTQ sequence; "
                          "only A, C, G, T and N are expected!");
    }
  }

  encoding.decode(m_qualities);
}
```

**Comment by MikkelSchubert** [comment]
Yeah, exactly. There and in `fastq::clean_sequence` in v2.3.x.

I'm currently busy preparing for a conference, but I'll try to release an update next week.

Best,
Mikkel

**Comment by MikkelSchubert** [comment]
Just to be sure that I don't release something that doesn't actually fit your needs:

Are we in agreement that ambiguous bases covers IUPAC codes [1] B, D, H, K, M, N, R, S, V, W, and Y?
I'd prefer not to just mask anything non-ACGTN, since that hides problems such as corrupted data.

And is support for U(racil) of interest? Those could conceivably be converted to T and processed as normal, possibly as a separate option from masking of ambiguous bases.

[1] https://en.wikipedia.org/wiki/Nucleic_acid_notation#IUPAC_notation

**Comment by 7PintsOfCherryGarcia** [comment]
> Are we in agreement that ambiguous bases covers IUPAC codes [1] B, D, H, K, M, N, R, S, V, W, and Y?

Correct, just need the support for IUPAC codes. And conversion to N works perfectly fine.

> And is support for U(racil) of interest? Those could conceivably be converted to T and processed as normal, possibly as a separate option from masking of ambiguous bases.

For sure!!

Thanks a lot!!

**Comment by MikkelSchubert** [comment]
Hi Julian,

I've just released [AdapterRemoval v3 alpha 2](https://github.com/MikkelSchubert/adapterremoval/releases/tag/v3.0.0-alpha2), which includes the new options `--mask-degenerate-bases` and `--convert-uracils`, that enable the operations we discussed. If you have the opportunity, then it would be great if you could give it a try and see if it works as you expect.

I'm attending a conference the next couple of days, for which I needed the alpha release to be out, but after that I'll port those options to AdapterRemoval v2, along with a couple of minor bug fixes, and release a new version by the end of the week.

Best,
Mikkel

**Comment by MikkelSchubert** [comment]
I've now released v2.3.4 which adds the new command-line options I mentioned above:
https://github.com/MikkelSchubert/adapterremoval/releases/tag/v2.3.4

I'd greatly appreciate it if you could give a spin and see if everything works as expected.

Best,
Mikkel
